lib/rubocop/cop/rspec/describe_class.rb

```diff
@@ -22,49 +22,38 @@ module RSpec
       #   describe "A feature example", type: :feature do
       #   end
       class DescribeClass < Base
-        include RuboCop::RSpec::TopLevelDescribe
+        include RuboCop::RSpec::TopLevelGroup
 
         MSG = 'The first argument to describe should be '\
               'the class or module being tested.'
 
-        def_node_matcher :valid_describe?, <<-PATTERN
-          {
-            (send #{RSPEC} :describe const ...)
-            (send #{RSPEC} :describe)
-          }
-        PATTERN
-
-        def_node_matcher :describe_with_rails_metadata?, <<-PATTERN
-          (send #{RSPEC} :describe !const ...
-            (hash <#rails_metadata? ...>)
-          )
-        PATTERN
-
         def_node_matcher :rails_metadata?, <<-PATTERN
           (pair
             (sym :type)
-            (sym {
-                   :channel :controller :helper :job :mailer :model :request
-                   :routing :view :feature :system :mailbox
-                 }
-            )
+            (sym { :channel :controller :helper :job :mailer :model :request
+                   :routing :view :feature :system :mailbox })
           )
         PATTERN
 
-        def on_top_level_describe(node, (described_value, _))
-          return if shared_group?(root_node)
-          return if valid_describe?(node)
-          return if describe_with_rails_metadata?(node)
-          return if string_constant_describe?(described_value)
+        def_node_matcher :not_a_const_described, <<~PATTERN
+          (block
+            [  (send #{RSPEC} :describe $[!const !#string_constant?] ...)
+              !(send #{RSPEC} :describe ... (hash <#rails_metadata? ...>)) ]
+            ...
+          )
+        PATTERN
```

Comment:
Collapsed several significantly overlapping node patterns into one.
Moved guard statements into node pattern as well.

There still seems to be some redundancy in this node pattern, similarly to `SubjectStub`:
```
          {
            (block (send nil? :subject (sym $_)) args ...)
            (block (send nil? $:subject) args ...)
          }
```

I'll open a ticket for `rubocop-ast`.

Reply:
Is `[ ...]` a new syntax for `{ ... }`? Does it work on the minimum supported rubocop version or should we bump the requirements?

Reply:
`[ ]` is logical "and", it's around for a while, just not too frequently used.

I find it too hard to reason about boolean login, this node pattern sounds like:
a block
  with a send of `describe` to `RSpec`/nil?
    where the first argument is
      neither a const not a string resembling a const
      *and* the rest of the arguments don't contain a hash with Rails-like metadata

Reply:
I'm confused by the `(send #rspec? :describe ` part in the second argument of `[ ]`

Reply:
@Darhazer Why?

It's basically the same as previously (which I forgot to remove during merge resolution):
```ruby
        def_node_matcher :describe_with_rails_metadata?, <<-PATTERN
          (send #rspec? :describe !const ...
            (hash <#rails_metadata? ...>)
          )
        PATTERN
```

`!const` check is not necessary anymore, it's checked in the first part of `[ ]`.

Reply:
I too find this node pattern a bit cryptic. Perhaps a bit more duplication can be justified if it makes this code easier to read?

Reply:
Good point guys. Split this node pattern into two parts. Looks more readable now.

Reply:
FYI the shiny newly released `rubocop-ast` allows comments in patterns